Eurosat

Prateek0xeo opened the pull request:
#104 
md5 mismatch issue solved

cregouby commented:
Hello @Prateek0xeo,

Thanks a lot for this contribution ! 

I have a few comment on it : 

- We do not want to host archive files in the code repository, as it would make the package too heavy, up to the point it would not pass the CRAN check. Could you remove those zip and tar.gz file from the pull-request please ?
-  the check fails with information that should be fixed : 
```log
==> devtools::check()

══ Documenting ══════════════════════════════════════════════════════════
ℹ Updating torchvision documentation
ℹ Loading torchvision
Avis : Objects listed as exports, but not present in namespace:
• inat2021_dataset
✖ dataset-eurosat.R:28: @export must be a single line, not 5.
ℹ The first line is "Download the EuroSAT dataset"
✖ transforms-magick.R:56: S3 method `get_image_size.magick-image` needs
  @export or @exportS3method tag.
Writing NAMESPACE
...
── R CMD check results ────────────────────── torchvision 0.6.0.9000 ────
Duration: 47.6s

❯ checking Rd \usage sections ... WARNING
  Documented arguments not in \usage in Rd file 'eurosat_dataset.Rd':
    ‘i’
  
  Functions with \usage entries need to have the appropriate \alias
  entries, and all their arguments documented.
  The \usage entries must correspond to syntactically valid R code.
  See chapter ‘Writing R documentation files’ in the ‘Writing R
  Extensions’ manual.
```
Could you please fix those two as well ?

Thanks a lot in advance.

Prateek0xeo commented:
Sure! Thank you for the feedback, I'll improve upon the PR and make sure that all other Pull requests i make also reflect the same.

cregouby commented:
Then checking the use of `ds <- eurosat_dataset(download = TRUE)` I get an SSL error from the server certificate: 
```log
utils::download.file(self$resources$url, destfile = zip_file,  :
  URL 'https://madm.dfki.de/files/sentinel/EuroSAT.zip': statut 'SSL peer certificate or SSH remote key was not OK'
```
Thus `libcurl` do not download the zip file and the dataloader fails.

Error is  confirmed by online SSL checkers : 
![image](https://github.com/user-attachments/assets/0f2b040d-44b9-4645-bff8-1e546b40712e)

cregouby commented:
I've contact the website owner for a fix / workaround.

Prateek0xeo commented:
@cregouby I have implemented all the changes you mentioned.
i have now removed the data files from the PR and improved Documents and ran the check cmd which gave no errors.
Pls check the PR once and tell me if there is anything else left unchecked so that i can move forward with rest of the Dataset loaders

Prateek0xeo commented:
Hello @cregouby,

I referred to the PyTorch repository and tried implementing the new URL and MD5 checksum provided in the code, but it still didn't resolve the issue. While the SSL certificate problem was fixed there, the dataset still doesn't load as expected.

I also took inspiration from the spam-loader review by @dfalbel and applied similar improvements to this repository.

cregouby commented:
Hello @Prateek0xeo 

I've recieved an answer from the website administrator: 
>  Dear Christophe, 
First of all, thank you for reaching out and for your interest in our dataset. 
Unfortunately, this is not the first time this issue has occurred, so I decided to host the dataset on Zenodo as well. 
For a permanent fix, I would recommend using the version hosted on Zenodo (https://zenodo.org/records/7711810). 
If you have any questions or need further information, please feel free to contact me.
Kind regards, 

I can see in the meantime that you switch to huggingface dataset, which fails in my case with a wrong MD5 
```
> ds <- eurosat_dataset(download = TRUE)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1173  100  1173    0     0   1123      0  0:00:01  0:00:01 --:--:--  1124
Erreur dans self$download() : 
  Downloaded file has an invalid MD5 checksum. Please try again.
Called from: self$download()
```
We can find the download failure root cause in the zip file : 
```r
Browse[1]> readLines(zip_file)
[1] "Found. Redirecting to https://cdn-lfs.hf.co/repos/fc/1d/fc1dee780dee1dae2ad48856d0961ac6aa5dfcaaaa4fb3561be4aedf19b7ccc7/8ebea626349354c5328b142b96d0430e647051f26efc2dc974c843f25ecf70bd?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27EuroSAT.zip%3B+filename%3D%22EuroSAT.zip%22%3B&response-content-type=application%2Fzip&Expires=1737474034&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTczNzQ3NDAzNH19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5oZi5jby9yZXBvcy9mYy8xZC9mYzFkZWU3ODBkZWUxZGFlMmFkNDg4NTZkMDk2MWFjNmFhNWRmY2FhYWE0ZmIzNTYxYmU0YWVkZjE5YjdjY2M3LzhlYmVhNjI2MzQ5MzU0YzUzMjhiMTQyYjk2ZDA0MzBlNjQ3MDUxZjI2ZWZjMmRjOTc0Yzg0M2YyNWVjZjcwYmQ%7EcmVzcG9uc2UtY29udGVudC1kaXNwb3NpdGlvbj0qJnJlc3BvbnNlLWNvbnRlbnQtdHlwZT0qIn1dfQ__&Signature=buJTsafUfhVxwvReQoUSJcNl3VwU74g1YghMxdzBwM6G25WA3-eyL3M6edcgNDBDCEBGjaoJ026dpGx9RZ%7EXsyirvlbOdND95c1jOqIZU5Y3Lh67bSvZcbbSbgGOjq5OvHvKbMSmXBWvPGzG3Ody8Z8Fm%7ExN9mhW6jqwi6LaQ84pH2VW-bWLLXTyrzexXmrmY2FBplfM3ZGD%7EmiTx1JHu97SbY7S8x9GYAPW16vV6w%7EVIWJ5KztnTpF%7Ea78q8tYWasUttVF9q9SbMw2DJMLq3s7MyFZlxXnO6cSM4jMN1fowTxbjbV5-O6rSpf0jp0-FtpK5DfVQWEvTAnBZ3ZOg%7Ew__&Key-Pair-Id=K3RPWS32NSSJCE"
```

So see my remarks inline.

cregouby commented on R/dataset-eurosat.R:
```diff
@@ -0,0 +1,113 @@
+      self$resources$url,
+      destfile = zip_file,
+      mode = "wb",
+      method = "curl",
```
**todo** **design** : default `libcurl` method do not require specific installation, and is following redirections, where `curl` don't. So I would keep the default

cregouby commented on R/dataset-eurosat.R:
```diff
@@ -0,0 +1,113 @@
+      destfile = zip_file,
+      mode = "wb",
+      method = "curl",
+      extra = "--insecure"
```
**todo** **security** : as mentionned in https://curl.se/docs/manpage.html#-k, this should not be used. Please remove.

cregouby commented on tests/testthat/test-dataset-eurosat.R:
```diff
@@ -0,0 +1,22 @@
+  expect_true(length(files) > 0, info = "Files should be downloaded in the temporary directory")
+  
+  extracted_dir <- file.path(temp_root, "2750")
+  expect_true(dir.exists(extracted_dir), info = "Extracted data folder should exist")
```
This fails to me : 
```r
>   expect_true(dir.exists(extracted_dir), info = "Extracted data folder should exist")
Erreur : dir.exists(extracted_dir) is not TRUE

`actual`:   FALSE
`expected`: TRUE 
Extracted data folder should exist
> head(files, 2)
[1] "/tmp/Rtmpmm3AfN/file6970544f1c340//eurosat/2750/AnnualCrop/AnnualCrop_1.jpg"  "/tmp/Rtmpmm3AfN/file6970544f1c340//eurosat/2750/AnnualCrop/AnnualCrop_10.jpg"

```

cregouby reviewed (changes requested):
Hello @Prateek0xeo 

Thanks for submitting, 
Please fix the remarks, and fix the tests.

**question** : as the huggingface dataset also includes a [train / test / val](https://huggingface.co/datasets/torchgeo/eurosat/blob/main/eurosat-test.txt) split, Is there a way to make those split those split available here ?

cregouby commented on R/dataset-eurosat.R:
```diff
@@ -0,0 +1,113 @@
+  name = "eurosat",
+  
+  resources = list(
+    url = "https://huggingface.co/datasets/torchgeo/eurosat/resolve/c877bcd43f099cd0196738f714544e355477f3fd/EuroSAT.zip",
```
**question** : The URL points to a specific commit of the Dataset onto huggingface platform. Is it what we want ? or do we want to use the latest version of this dataset ? In that case url would be the one accessible on the front-page of the Dataset:  `url = "https://huggingface.co/datasets/torchgeo/eurosat/resolve/main/EuroSAT.zip?download=true"`

Prateek0xeo replied:
https://github.com/pytorch/vision/pull/8563
I tried to implement the fix taking this issue in pytorch as reference. So i copied the same URL from the code. I'll implement the latest version of dataset as you pointed out.

Prateek0xeo commented:
> source("C:/Coding/torchvision/torchvision/R/dataset-eurosat.R")
> ds <- eurosat_dataset(download = TRUE)
trying URL 'https://huggingface.co/datasets/torchgeo/eurosat/resolve/main/EuroSAT.zip?download=true'
Content type 'application/zip' length 94280567 bytes (89.9 MB)
downloaded 89.9 MB

> sample <- ds[1]
> print(sample$x)
, , 1

           [,1]      [,2]      [,3]      [,4]      [,5]      [,6]      [,7]
 [1,] 0.5843137 0.5764706 0.5725490 0.5764706 0.5843137 0.5921569 0.5843137
 [2,] 0.5843137 0.5843137 0.5843137 0.5803922 0.5882353 0.5921569 0.5803922
 
 the dataloader seems to be working now I'll push the code after self review

Prateek0xeo commented:
@cregouby So i have pushed the code and everything seems fine i have run the test file all test cases passed. I also ran the ds <- eurosat_dataset(download = TRUE) cmd which downloaded the dataset. i have ran the devtools::check cmd and everthing seems perfect. If any issue still persists pls do comment.

Now as for the minst_dataset() I'll make a separate PR to implement the [train / test / val] split

Prateek0xeo commented:
@cregouby In this commit dataloader is working. I am still working on figuring out the splits functionalilty until then pls review the new commit.

Prateek0xeo commented:
Hello @cregouby 
I contacted the dataset owner and received this message

Hi Prateek,

All of the datasets in https://huggingface.co/datasets/torchgeo/ are intended for use with the TorchGeo library. You can use it like so:

from torchgeo.datasets import EuroSAT

train_dataset = EuroSAT('data', split='train', download=True)
val_dataset = EuroSAT('data', split='val', download=True)
test_dataset = EuroSAT('data', split='test', download=True)
See https://torchgeo.readthedocs.io/en/latest/api/datasets.html#eurosat for documentation.

So i implemented the functionality using these URL Endpoints
List the split names for this dataset


curl -X GET \
     ["https://datasets-server.huggingface.co/splits?dataset=torchgeo%2Feurosat"](https://datasets-server.huggingface.co/splits?dataset=torchgeo%2Feurosat)
     
but the Val split was not working and throwing an error
<img width="858" alt="Screenshot 2025-01-20 210917" src="https://github.com/user-attachments/assets/478ff629-7b14-46cb-872b-25f4ec15595b" />

So i have contacted the Owner again for this issue
The new PR should work once the dataset host is fixed
Pls go through the new commit and comment for any change

Prateek0xeo commented:
@cregouby  there is a problem with image decoding with this dataset 
ERROR - The dataset API returned a value for row$image that is not a valid Base64-encoded string, and the decoding step failed. This issue likely arises from the structure or format of the dataset returned by the API
Because of this i can't access the image.

cregouby commented:
Hello @Prateek0xeo 

1 - The new download only gets the first 100 sample, which is not what we want. We want the full train dataset, i.e. 32.5k rows.
2 - You should not **replace** the previous test, but rather add the new tests in the same file.

I get the following error running the tests : 
```
[ FAIL 2 | WARN 1 | SKIP 0 | PASS 3 ]

── Warning (test-dataset-eurosat.R:18:3): eurosat_dataset API handles splits correctly via API ─────────────────────────────────────────────────────────────────────────────────────────────
Skipping row 1 due to image decoding failure: I can only decode base64 strings
Backtrace:
    ▆
 1. ├─train_ds[1] at test-dataset-eurosat.R:18:3
 2. └─torch:::`[.dataset`(train_ds, 1) at test-dataset-eurosat.R:18:3
 3.   └─x$.getitem(y)
 4.     └─base::tryCatch(...) at torchvision_prateek/R/dataset-eurosat.R:82:5
 5.       └─base (local) tryCatchList(expr, classes, parentenv, handlers)
 6.         └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
 7.           └─value[[3L]](cond)

── Failure (test-dataset-eurosat.R:19:3): eurosat_dataset API handles splits correctly via API ─────────────────────────────────────────────────────────────────────────────────────────────
!is.null(sample$x) is not TRUE

`actual`:   FALSE
`expected`: TRUE 
Image should not be null

── Failure (test-dataset-eurosat.R:20:3): eurosat_dataset API handles splits correctly via API ─────────────────────────────────────────────────────────────────────────────────────────────
!is.null(sample$y) is not TRUE

`actual`:   FALSE
`expected`: TRUE 
Label should not be null
[ FAIL 2 | WARN 1 | SKIP 0 | PASS 3 ]

```

In order to ensure the test runs correctly on all platforms, you should activate the github actions in the setup of your git repository, and then activate the R-CMD-check workflow. 
![image](https://github.com/user-attachments/assets/4590d80d-ca05-47b9-a2fc-9a27b5cdc647)


So that for each or your commit, you will see test results over all platforms.  All must be green, or you have to fix code.

Prateek0xeo commented:
@cregouby Should i merge the branch add-Eurosat-dataloader with main and then conduct workflow run or can I R-CMD-check.yaml to the branch?

cregouby commented:
Hello @Prateek0xeo 
There is no need to change something in th e`.github/workflow/` folder as everything is already here. 
The setup I mentionned is to be done in **your** github repository.

Prateek0xeo commented:
@cregouby yes actually I was trying figure out that only when I noticed it
was already done in the git repo so I deleted the folder.
Update - I have made the split work but the problem is I can't download
more than 100 from the URL. I tried using parquet but that also failed. I
am now trying to use the text file to split the data downloaded from the zip

On Thu, 23 Jan, 2025, 00:27 cregouby, ***@***.***> wrote:

> Hello @Prateek0xeo <https://github.com/Prateek0xeo>
> There is no need to change something in th e.github/workflow/ folder as
> everything is already here.
> The setup I mentionned is to be done in *your* github repository.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mlverse/torchvision/pull/122#issuecomment-2608030121>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BAG3KFCC2OWZICS2BENR7M32L7S23AVCNFSM6AAAAABVAE7UOGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBYGAZTAMJSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cregouby commented on tests/testthat/test-dataset-eurosat.R:
```diff
@@ -0,0 +1,21 @@
+  
+  # Test train split
+  train_ds <- eurosat_dataset(root = "./data/eurosat", split = "train", download = TRUE)
+  expect_true(length(train_ds) > 0, info = "Train dataset should have a non-zero length")
```
```suggestion
  expect_length(train_ds), 16200, info = "Train dataset should have the expected length")
```

cregouby commented on tests/testthat/test-dataset-eurosat.R:
```diff
@@ -0,0 +1,21 @@
+  
+  # Test test split
+  test_ds <- eurosat_dataset(root = "./data/eurosat", split = "test", download = TRUE)
+  expect_true(length(test_ds) > 0, info = "Test dataset should have a non-zero length")
```
```suggestion
  expect_length(test_ds), 5400,, info = "Test dataset should have the expected  length")
```

cregouby commented on tests/testthat/test-dataset-eurosat.R:
```diff
@@ -0,0 +1,21 @@
+  
+  # Test validation split
+  validation_ds <- eurosat_dataset(root = "./data/eurosat", split = "validation", download = TRUE)
+  expect_true(length(validation_ds) > 0, info = "Validation dataset should have a non-zero length")
```
```suggestion
  expect_length(validation_ds), 5400, info = "Validation dataset should have the expected length")
```

cregouby commented on tests/testthat/test-dataset-eurosat.R:
```diff
@@ -0,0 +1,21 @@
+  
+  # Test data content
+  sample <- train_ds[1]
+  expect_true(!is.null(sample$x), info = "Image should not be null")
+  expect_true(!is.null(sample$y), info = "Label should not be null")
```
**todo** : would be much better to check the **tensor shape** and **tensor dtype** of each x and y object

Prateek0xeo commented:
@cregouby all R-CMD-check are passed

cregouby commented:
Hello @Prateek0xeo 

Can you review my 5 suggested changes  ? https://github.com/mlverse/torchvision/pull/122/files/de833e1abb1976fa7af915c6feb490c06128785e

Prateek0xeo reviewed (commented):
suggestions are implemented
1. the text file data is length is checked for exact amount of rows (3 suggestions)
2.check added for tensor shape and tensor dtype of each x and y object (1 suggestion)
What is the 5th suggestion? @cregouby

cregouby commented:
Hello @Prateek0xeo 

Tests fails for me : 
```r
[ FAIL 2 | WARN 0 | SKIP 0 | PASS 5 ]

── Failure (test-dataset-eurosat.R:26:3): eurosat_dataset API handles splits correctly via API ──────────────────────────────────────────────────────────────────────────────────────────────
Train dataset should have exactly 16200 samples not equal to 16200.
1/1 mismatches
[1] 100 - 16200 == -16100

── Error (test-dataset-eurosat.R:28:3): eurosat_dataset API handles splits correctly via API ────────────────────────────────────────────────────────────────────────────────────────────────
Error in `initialize(...)`: Invalid split. Choose one of 'train', 'validation', or 'test'.
Backtrace:
    ▆
 1. └─torch (local) eurosat_dataset(root = temp_root, split = "val", download = TRUE) at test-dataset-eurosat.R:28:3
 2.   └─d$new(...)
 3.     └─torchvision (local) initialize(...)
[ FAIL 2 | WARN 0 | SKIP 0 | PASS 5 ]
```
but it should'd be that hard to fix...

For readability, I would recommand to use the most advanced testthat capability. I.e. prefer `expect_length()` over `expect_true(length())` and `expect_match()` instead of `expect_true(!is.null()) and (expect_equal(type_of(), "character")`.
Please note that we never expect `typeof() == ` as type are additive, but rather `expect_is()` which test for inheritance.

I would recommand you to mimic the tests that are performed on MNIST in  `tests/testthat/test-dataset-mnist.R` L3-L28 for a good coverage (and specifically the ability to get torch_tensors out of `transform = transform_to_tensor` )

cregouby commented:
Hello @Prateek0xeo 

Trying to update the FR translation with the new messages related to your work, I realize I can't. Would you mind grant me access to your github repo ?

Thanks

Prateek0xeo commented:
Yeah sure

On Thu, 30 Jan, 2025, 13:36 cregouby, ***@***.***> wrote:

> Hello @Prateek0xeo <https://github.com/Prateek0xeo>
>
> Trying to update the FR translation with the new messages related to your
> work, I realize I can't. Would you mind grant me access to your github repo
> ?
>
> Thanks
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mlverse/torchvision/pull/122#issuecomment-2623794868>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BAG3KFHJ2QBN6WQVUNWPCTT2NHMPZAVCNFSM6AAAAABVAE7UOGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMRTG44TIOBWHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cregouby commented:
Hello @Prateek0xeo,

I've updated the test to check that a dataloader can feed a model with torch tensor. And therre is a problem with `y` not being a tensor but a character (that will be unusable with torch).

`y` must be a `torch_long()` as per the test expectation. 

You can achieve this the way it is done for cifar and add a mapping function to labels like in https://github.com/mlverse/torchvision/blob/df06f28be489830633a870b5d0295acd57865987/R/dataset-cifar.R#L58-L62

Prateek0xeo commented:
Thank you for your help @cregouby , is there anymore changes required or is this PR complete?

cregouby commented:
Hello @Prateek0xeo ,

Sorry, I was not clear enough in my message : 
- Please implement a `.load_meta` -like method for the classes to be available in the dataset object
- Please turn the y vector into a torch_long() object
- Then you could add a entry in the NEWS file for the new available dataset, and add your name as contributor ('ctb') in the DESCRIPTION file.
Then I guess we'll be able to merge it !
Great job anyway, I can see the learning curve through your contributions.

Prateek0xeo commented:
Thanks a lot for the praise, @cregouby! I'm currently busy with exams, but I'll try to push the change by next week. Hopefully, I can complete my first PR to torchvision soon!

Prateek0xeo commented:
@cregouby  So ,
1. i have added a load_meta in the dataloader
`if (!file.exists(self$split_file)) {
      stop(sprintf("Split file not found for split='%s'.", split))
    }
    
    self$data <- suppressWarnings(readLines(self$split_file))
    self$load_meta()
  },
  
  load_meta = function() {
    self$classes <- unique(sub("_.*", "", self$data))
    self$class_to_idx <- setNames(seq_along(self$classes) - 1, self$classes)
  },`
 2. i did turn the y into torch long but had a issue
 y is not being returned as a scalar tensor but as a tensor with shape (1,), which causes the shape mismatch in expect_tensor_shape(i[[2]], 10)
 so i used squeeze () to remove dimensions of size 1
 then i did an RMD check 
 which gave me a character problem with label
so i modified the getitem code to this
`.getitem = function(index) {
    filename <- self$data[index]
    label <- as.character(sub("_.*", "", filename))  # Ensure label is a character string
    
    image_path <- file.path(self$images_dir, "2750", label, filename)
    if (!file.exists(image_path)) {
      stop("Image file not found: ", image_path)
    }
    
    img_array <- jpeg::readJPEG(image_path)
    
    if (!is.null(self$transform)) {
      img_array <- self$transform(img_array)
    }
    
    # Ensure label exists in class_to_idx
    if (!label %in% names(self$class_to_idx)) {
      stop("Label not found in class_to_idx: ", label)
    }
    
    # Convert label index to torch tensor with dtype = torch_long()
    label_idx <- torch::torch_tensor(as.integer(self$class_to_idx[[label]]), dtype = torch_long())$squeeze()
    
    list(x = img_array, y = label_idx)
  }`
  after this all RMD checks passed and the dataloader is working
  `> sample <- ds[34]
> print(sample$y)
torch_tensor
2
[ CPULongType{} ]
`
3. I have added my name as contributer on the description file

cregouby commented:
Hello @Prateek0xeo,

There is a mismatch between the documentation for param `split` and the code managing it. Could you please fix that ?

Prateek0xeo commented:
@cregouby I have fixed the document to this
@param split Character. One of `train`, `val`, or `test`

i also tested using val in split which is working
<img width="495" alt="image" src="https://github.com/user-attachments/assets/4d4062aa-d4a3-4ad4-9adb-3d4129e1e689" />

cregouby reviewed (approved):
**praise** : thanks for adding this dataset. Porting huggging-face datasets opens the door to a lot of other datasets in {torchvision}
**style** **non-blocking** : a lot of style and design refactoring will take place later-on for consistency with other `_dataset()` functions (I'll raise an issue for this)

Prateek0xeo commented:
@cregouby 
so i have replaced sprintf with glue::glue , used runtime package and copied messages from the other dataset loader but improvised the messages as per requirement.

cregouby commented:
superseeded by #126
